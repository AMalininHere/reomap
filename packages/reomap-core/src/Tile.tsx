import React, { useState, Ref, useCallback, useEffect, useRef } from 'react';

type Props = React.ImgHTMLAttributes<HTMLImageElement>;

function Tile({ style, onLoad, src, ...props }: Props, ref: Ref<HTMLImageElement>) {
  const [ loaded, setLoaded ] = useState(false);

  const prevSrc = useRef(src);

  useEffect(() => {
    if (prevSrc.current !== src) {
      setLoaded(false);
    }

    prevSrc.current = src;
  }, [src]);

  const handleLoad = useCallback<Required<Props>['onLoad']>(e => {
    setLoaded(true);
    return onLoad?.(e);
  }, [ onLoad ]);

  if (!loaded) {
    style = {
      ...style,
      visibility: 'hidden'
    };
  }

  return (
    <img ref={ref} style={style} onLoad={handleLoad} src={src} {...props}/>
  );
}

export default React.forwardRef(Tile);
