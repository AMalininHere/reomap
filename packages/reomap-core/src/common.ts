import { tile2lat, tile2lng, lng2tile, lat2tile } from './utils/geo-fns';

export class LatLng {
  constructor(
    public readonly lat: number,
    public readonly lng: number,
  ) { }

  equals(other: LatLng) {
    return (
      this.lat === other.lat &&
      this.lng === other.lng
    );
  }
}

export class Point {
  constructor(
    public readonly x: number,
    public readonly y: number,
  ) { }
}

export const TILE_SIZE = 256;

export const absMinLatLng = new LatLng(
  tile2lat(Math.pow(2, 10), 10),
  tile2lng(0, 10)
);

export const absMaxLatLng = new LatLng(
  tile2lat(0, 10),
  tile2lng(Math.pow(2, 10), 10)
);

export function pixelToLatLng(width: number, height: number, zoom: number, center: LatLng, source: Point) {
  const pointDiffX = (source.x - width / 2) / TILE_SIZE;
  const pointDiffY = (source.y - height / 2) / TILE_SIZE;

  const tileX = lng2tile(center.lng, zoom) + pointDiffX;
  const tileY = lat2tile(center.lat, zoom) + pointDiffY;

  return new LatLng(
    Math.max(absMinLatLng.lat, Math.min(absMaxLatLng.lat, tile2lat(tileY, zoom))),
    Math.max(absMinLatLng.lng, Math.min(absMaxLatLng.lng, tile2lng(tileX, zoom)))
  );
}

export function latLngToPixel(width: number, height: number, zoom: number, center: LatLng, source: LatLng) {
  const tileCenterX = lng2tile(center.lng, zoom);
  const tileCenterY = lat2tile(center.lat, zoom);

  const tileX = lng2tile(source.lng, zoom);
  const tileY = lat2tile(source.lat, zoom);

  return new Point(
    (tileX - tileCenterX) * TILE_SIZE + width / 2,
    (tileY - tileCenterY) * TILE_SIZE + height / 2
  );
}
