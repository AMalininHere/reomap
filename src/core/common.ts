import { tile2lat, tile2lng, lng2tile, lat2tile } from './utils/geo-fns';
import { LatLng, Point } from './models';

export const absMinLatLng = new LatLng(
  tile2lat(Math.pow(2, 10), 10),
  tile2lng(0, 10)
);

export const absMaxLatLng = new LatLng(
  tile2lat(0, 10),
  tile2lng(Math.pow(2, 10), 10)
);

export function pixelToLatLng(width: number, height: number, zoom: number, center: LatLng, source: Point) {
  const pointDiffX = (source.x - width / 2) / 256.0;
  const pointDiffY = (source.y - height / 2) / 256.0;

  const tileX = lng2tile(center.lng, zoom) + pointDiffX;
  const tileY = lat2tile(center.lat, zoom) + pointDiffY;

  return new LatLng(
    Math.max(absMinLatLng.lat, Math.min(absMaxLatLng.lat, tile2lat(tileY, zoom))),
    Math.max(absMinLatLng.lng, Math.min(absMaxLatLng.lng, tile2lng(tileX, zoom)))
  );
}

export function latLngToPixel(width: number, height: number, zoom: number, center: LatLng, source: LatLng) {
  const tileCenterX = lng2tile(center.lng, zoom);
  const tileCenterY = lat2tile(center.lat, zoom);

  const tileX = lng2tile(source.lng, zoom);
  const tileY = lat2tile(source.lat, zoom);

  return new Point(
    (tileX - tileCenterX) * 256.0 + width / 2,
    (tileY - tileCenterY) * 256.0 + height / 2
  );
};
