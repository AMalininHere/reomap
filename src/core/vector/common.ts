import * as G from 'geojson';
import { LatLng, Point } from '../models';

export interface ElementProps<T extends G.Geometry> {
  geoElement: T;
  latLngToPixel: (ll: LatLng) => Point;
}