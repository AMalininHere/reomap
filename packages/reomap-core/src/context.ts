import React, { useContext } from 'react';
import { LatLng, Point, latLngToPixel, pixelToLatLng } from './common';

export class ContextState {
  constructor(
    public readonly center: LatLng,
    public readonly zoom: number,
    public readonly width: number,
    public readonly height: number,
  ) { }

  public readonly latLngToPixel = (source: LatLng) =>
    latLngToPixel(this.width, this.height, this.zoom, this.center, source);

  public readonly pixelToLatLng = (source: Point) =>
    pixelToLatLng(this.width, this.height, this.zoom, this.center, source);
}

const ctx = React.createContext<ContextState>(null!);

export const MapProvider = ctx.Provider;

export function useMapContext() {
  const data = useContext(ctx);

  return data;
}
