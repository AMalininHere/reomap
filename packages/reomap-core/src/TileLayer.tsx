import { lng2tile, lat2tile } from './utils/geo-fns';
import { useMapContext } from './context';
import { TILE_SIZE } from './common';
import Tile from './Tile';

function useTileValues() {
  const {
    center,
    zoom,
    width,
    height,
  } = useMapContext();


  const roundedZoom = Math.round(zoom);
  const zoomDiff = zoom - roundedZoom;

  const scale = Math.pow(2, zoomDiff);
  const scaleWidth = width / scale;
  const scaleHeight = height / scale;

  const tileCenterX = lng2tile(center.lng, roundedZoom);
  const tileCenterY = lat2tile(center.lat, roundedZoom);

  const halfWidth = scaleWidth / 2 / TILE_SIZE;
  const halfHeight = scaleHeight / 2 / TILE_SIZE;

  const tileMinX = Math.floor(tileCenterX - halfWidth);
  const tileMaxX = Math.floor(tileCenterX + halfWidth);

  const tileMinY = Math.floor(tileCenterY - halfHeight);
  const tileMaxY = Math.floor(tileCenterY + halfHeight);

  return {
    tileCenterX,
    tileCenterY,

    tileMinX,
    tileMaxX,

    tileMinY,
    tileMaxY,

    roundedZoom,

    scale,
    scaleWidth,
    scaleHeight,
  };
}

type TileProvider = (x: number, y: number, z: number) => string;

interface Props {
  provider: TileProvider;
  tileClassName?: string;
  className?: string;
}

function TileLayer(props: Props) {
  const {
    tileCenterX,
    tileCenterY,

    tileMinX,
    tileMaxX,

    tileMinY,
    tileMaxY,

    roundedZoom,

    scale,
    scaleWidth,
    scaleHeight,
  } = useTileValues();

  const xMin = Math.max(tileMinX, 0);
  const yMin = Math.max(tileMinY, 0);
  const xMax = Math.min(tileMaxX, Math.pow(2, roundedZoom) - 1);
  const yMax = Math.min(tileMaxY, Math.pow(2, roundedZoom) - 1);

  const { provider, tileClassName, className } = props;
  const tiles: React.ReactNode[] = [];

  for (let x = xMin; x <= xMax; ++x) {
    for (let y = yMin; y <= yMax; ++y) {
      tiles.push(
        <Tile
          key={`${x}-${y}-${roundedZoom}`}
          alt=""
          src={provider(x, y, roundedZoom)}
          loading="lazy"
          className={tileClassName}
          style={{
            position: 'absolute',
            width: TILE_SIZE,
            height: TILE_SIZE,
            left: (x - tileMinX) * TILE_SIZE,
            top: (y - tileMinY) * TILE_SIZE
          }}
        />
      );
    }
  }

  const left = -((tileCenterX - tileMinX) * TILE_SIZE - scaleWidth / 2);
  const top = -((tileCenterY - tileMinY) * TILE_SIZE - scaleHeight / 2);

  const tilesStyle: React.CSSProperties = {
    position: 'absolute',
    width: (tileMaxX - tileMinX + 1) * TILE_SIZE,
    height: (tileMaxY - tileMinY + 1) * TILE_SIZE,
    willChange: 'transform',
    transform: `translate(${left * scale}px, ${top * scale}px) scale(${scale})`,
    transformOrigin: 'top left',
  };

  return (
    <div className={className} style={tilesStyle}>
      {tiles}
    </div>
  );
}

export default TileLayer;
