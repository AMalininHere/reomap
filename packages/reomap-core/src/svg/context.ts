import { createContext, useContext } from 'react';
import { ContextState } from '../context';

const ctx = createContext<ContextState>(null!);
ctx.displayName = 'SvgLayerContext';

export const SvgLayerProvider = ctx.Provider;

export function useSvgLayerContext() {
  const data = useContext(ctx);

  return data;
}
