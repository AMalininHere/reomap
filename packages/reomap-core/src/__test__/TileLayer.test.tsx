import { ComponentProps } from 'react';
import { render } from '@testing-library/react';

import TileLayer from '../TileLayer';
import Tile from '../Tile';

import { MapProvider, createContextState } from '../context';
import { latLng } from '../common';

jest.mock('../Tile', () => (props: ComponentProps<typeof Tile>) => <img {...props}/>);

describe('TileLayer', () => {
  const tileProvider: ComponentProps<typeof TileLayer>['provider'] = (x, y, z) => `tile-${x}-${y}-${z}`;

  it('should render tiles', () => {

    const contextState = createContextState(latLng(55.417, 85.276), 6, 800, 600);

    const { container } = render(
      <MapProvider value={contextState}>
        <TileLayer provider={tileProvider}/>
      </MapProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
