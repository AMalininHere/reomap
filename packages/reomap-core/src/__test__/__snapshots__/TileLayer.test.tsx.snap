// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TileLayer should render tiles 1`] = `
<div>
  <div
    style="position: absolute; width: 1024px; height: 1024px; will-change: transform; transform: translate(-153.00551111111236px, -240.9634059561813px) scale(1); transform-origin: top left;"
  >
    <img
      alt=""
      loading="lazy"
      src="tile-45-18-6"
      style="position: absolute; width: 256px; height: 256px; left: 0px; top: 0px;"
    />
    <img
      alt=""
      loading="lazy"
      src="tile-45-19-6"
      style="position: absolute; width: 256px; height: 256px; left: 0px; top: 256px;"
    />
    <img
      alt=""
      loading="lazy"
      src="tile-45-20-6"
      style="position: absolute; width: 256px; height: 256px; left: 0px; top: 512px;"
    />
    <img
      alt=""
      loading="lazy"
      src="tile-45-21-6"
      style="position: absolute; width: 256px; height: 256px; left: 0px; top: 768px;"
    />
    <img
      alt=""
      loading="lazy"
      src="tile-46-18-6"
      style="position: absolute; width: 256px; height: 256px; left: 256px; top: 0px;"
    />
    <img
      alt=""
      loading="lazy"
      src="tile-46-19-6"
      style="position: absolute; width: 256px; height: 256px; left: 256px; top: 256px;"
    />
    <img
      alt=""
      loading="lazy"
      src="tile-46-20-6"
      style="position: absolute; width: 256px; height: 256px; left: 256px; top: 512px;"
    />
    <img
      alt=""
      loading="lazy"
      src="tile-46-21-6"
      style="position: absolute; width: 256px; height: 256px; left: 256px; top: 768px;"
    />
    <img
      alt=""
      loading="lazy"
      src="tile-47-18-6"
      style="position: absolute; width: 256px; height: 256px; left: 512px; top: 0px;"
    />
    <img
      alt=""
      loading="lazy"
      src="tile-47-19-6"
      style="position: absolute; width: 256px; height: 256px; left: 512px; top: 256px;"
    />
    <img
      alt=""
      loading="lazy"
      src="tile-47-20-6"
      style="position: absolute; width: 256px; height: 256px; left: 512px; top: 512px;"
    />
    <img
      alt=""
      loading="lazy"
      src="tile-47-21-6"
      style="position: absolute; width: 256px; height: 256px; left: 512px; top: 768px;"
    />
    <img
      alt=""
      loading="lazy"
      src="tile-48-18-6"
      style="position: absolute; width: 256px; height: 256px; left: 768px; top: 0px;"
    />
    <img
      alt=""
      loading="lazy"
      src="tile-48-19-6"
      style="position: absolute; width: 256px; height: 256px; left: 768px; top: 256px;"
    />
    <img
      alt=""
      loading="lazy"
      src="tile-48-20-6"
      style="position: absolute; width: 256px; height: 256px; left: 768px; top: 512px;"
    />
    <img
      alt=""
      loading="lazy"
      src="tile-48-21-6"
      style="position: absolute; width: 256px; height: 256px; left: 768px; top: 768px;"
    />
  </div>
</div>
`;
